pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -18,158 +18,207 @@ package credentials
 
 import (
 	"encoding/base64"
+	"errors"
 	"fmt"
+	"net/url"
 	"strings"
 	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/request"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/ecr"
-	"k8s.io/klog"
 
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
 	"k8s.io/kubernetes/pkg/credentialprovider"
 	"k8s.io/kubernetes/pkg/version"
 )
 
-const awsChinaRegionPrefix = "cn-"
-const awsStandardDNSSuffix = "amazonaws.com"
-const awsChinaDNSSuffix = "amazonaws.com.cn"
-const registryURLTemplate = "*.dkr.ecr.%s.%s"
+// init registers a credential provider for each registryURLTemplate and creates
+// an ECR token getter factory with a new cache to store token getters
+func init() {
+	credentialprovider.RegisterCredentialProvider("amazon-ecr",
+		newECRProvider(&ecrTokenGetterFactory{cache: make(map[string]tokenGetter)}))
+}
 
-// awsHandlerLogger is a handler that logs all AWS SDK requests
-// Copied from pkg/cloudprovider/providers/aws/log_handler.go
-func awsHandlerLogger(req *request.Request) {
-	service := req.ClientInfo.ServiceName
-	region := req.Config.Region
+// ecrProvider is a DockerConfigProvider that gets and refreshes tokens
+// from AWS to access ECR.
+type ecrProvider struct {
+	cache         cache.Store
+	getterFactory tokenGetterFactory
+}
 
-	name := "?"
-	if req.Operation != nil {
-		name = req.Operation.Name
-	}
+var _ credentialprovider.DockerConfigProvider = &ecrProvider{}
 
-	klog.V(3).Infof("AWS request: %s:%s in %s", service, name, *region)
+func newECRProvider(getterFactory tokenGetterFactory) *ecrProvider {
+	return &ecrProvider{
+		cache:         cache.NewExpirationStore(stringKeyFunc, &ecrExpirationPolicy{}),
+		getterFactory: getterFactory,
+	}
 }
 
-// An interface for testing purposes.
-type tokenGetter interface {
-	GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput) (*ecr.GetAuthorizationTokenOutput, error)
+// Enabled implements DockerConfigProvider.Enabled. Enabled is true if AWS
+// credentials are found.
+func (p *ecrProvider) Enabled() bool {
```

Comment:
this logs errors in all cases that don't return true... does that mean the kubelet log will be spammed with "while validating AWS credentials" errors on every image pull in non-aws environments?

Reply:
Good point. Those should be a log of some level.

Reply:
if this runs in a non-aws environment, I wouldn't really expect any logging

Reply:
Yes, GCP and Azure do. And it's not so much not being in their effective environment as opposed to not being able to enable to provider. The extent of AWS environment is having AWS credentials.

GCP
```
// Returns true if it finds a local GCE VM.
// Looks at a product file that is an undocumented API.
func onGCEVM() bool {
	data, err := ioutil.ReadFile(gceProductNameFile)
	if err != nil {
		klog.V(2).Infof("Error while reading product_name: %v", err)
		return false
	}
	name := strings.TrimSpace(string(data))
	return name == "Google" || name == "Google Compute Engine"
}

// Enabled implements DockerConfigProvider for all of the Google implementations.
func (g *metadataProvider) Enabled() bool {
	return onGCEVM()
}
```
Azure
```
func (a *acrProvider) Enabled() bool {
	if a.file == nil || len(*a.file) == 0 {
		klog.V(5).Infof("Azure config unspecified, disabling")
		return false
	}

	f, err := os.Open(*a.file)
	if err != nil {
		klog.Errorf("Failed to load config from file: %s", *a.file)
		return false
	}
	defer f.Close()

	err = a.loadConfig(f)
	if err != nil {
		klog.Errorf("Failed to load config from file: %s", *a.file)
		return false
	}

	a.servicePrincipalToken, err = auth.GetServicePrincipalToken(a.config, a.environment)
	if err != nil {
		klog.Errorf("Failed to create service principal token: %v", err)
		return false
	}

	a.registryClient = newAzRegistriesClient(a.config.SubscriptionID, a.environment.ResourceManagerEndpoint, a.servicePrincipalToken)
	return true
}
```

Reply:
> The extent of AWS environment is having AWS credentials.

:+1: You are not required to run in EC2 in order to use ECR, you just need to have valid AWS credentials.

Reply:
> Yes, GCP and Azure do. And it's not so much not being in their effective environment as opposed to not being able to enable to provider. The extent of AWS environment is having AWS credentials.

ok